[DNM] soc: arm: cypress: psoc6: Introduce pinctrl

nandojve opened the pull request:
Introduce pinctrl capability on Cypress PSoC-6 SoC.  This add SoC support to configure port pins with flags using device tree.
Add myself as codeowner since PSoC-6 is orphan

This is a split from #28527.
It can be merged only after #28779.

Signed-off-by: Gerson Fernando Budke <gerson.budke@atl-electronics.com>

nandojve committed:
soc: arm: cypress: psoc6: Introduce pinctrl

Introduce pinctrl capability on Cypress PSoC-6 SoC.  This add SoC
support to configure port pins with flags using device tree.

Signed-off-by: Gerson Fernando Budke <gerson.budke@atl-electronics.com>

galak commented on dts/bindings/gpio/cypress,psoc6-gpio.yaml:
```diff
@@ -0,0 +1,37 @@
+  - port
+  - pin
+  - peripheral
+  - flags
```
we should handle flags like we are doing on atmel as explicit properties

nandojve replied:
Done!

```
#define DT_CYPRESS_HSIOM_FLAG(flag) flag;
#define DT_CYPRESS_HSIOM_FLAGS(...) \
	MACRO_MAP_CAT(DT_CYPRESS_HSIOM_FLAG __VA_OPT__(,) __VA_ARGS__)

#define DT_CYPRESS_HSIOM(inst, signal, port, pin, hsiom, ...) \
	p##port##_##pin##_##inst##_##signal: \
	p##port##_##pin##_##inst##_##signal { \
	cypress,pins = < &p##port port pin HSIOM_SEL_##hsiom > ; \
		DT_CYPRESS_HSIOM_FLAGS(__VA_ARGS__) \
	}
```
Now it have a variable length flags list. For an entry like this `DT_CYPRESS_PIN(uart5, rx, 5, 0, act_4, bias-pull-up, input-buffer);` It will construct as:
```
			p5_0_uart5_rx: p5_0_uart5_rx {
				cypress,pins = < &p5 0x5 0x0 0x12 >;
				bias-pull-up;
				input-buffer;
			};
```
The flags list can be zero length.

Drawback was, I needed to copy this https://github.com/zephyrproject-rtos/zephyr/pull/28645/commits/1f0c490272f107f8c8f6324d9b5812ff99d7bb9f#diff-e9536070e740fa654129a980ece66553R89-R98 from <sys/util.h> and add <sys/util_internal.h> to make it works. When include <sys/util.h> I got many path errors and garbage at **./build/zephyr/\<board\>.dts.pre.tmp**.

I send an update to show the whole solution and would like know how solve the copied content thing. Since <sys/util_internal.h> I should not use it (I think) and the content copied can not be on the <sys/util_internal.h> I guess.

nandojve replied:
Proposal fix 0eac735dc23cda700d53984bffd32894b18ab3d9

galak commented on dts/bindings/pinctrl/cypress,psoc6-pinctrl.yaml:
```diff
@@ -0,0 +1,59 @@
+      "cypress,pins":
+        type: phandle-array
+
+      bias-pull-up:
```
Can you see if you can pull the common properties here out into a pinctrl-flags.yaml and see if we can include that both here and in `atmel,sam-pinctrl.yaml`

nandojve replied:
Done at 05ba19e84abd1830a78ab60ed519a63711d5744a

nandojve committed:
include: sys: util: Move macros from util to util_macro

To elaborate more advanced macros for dts related bindings the
<sys/util.h> can not be include.  It create garbage at
<board>.dts.pre.tmp file and have path include issues.  This
moves all macros from util.h to util_macro.h  This allows that
<sys/util_macro.h> can be include on files like
<dts/arm/<manufacturer>/pinctrl_<manufacturer>_<soc>.h.

The issue was raised when try create a macro for pincrtl with
a flag list with variable length.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
dts: pinctrl: Introduce pincfg-node

Many data items that are represented in a pin configuration node are
common and generic. Pin control bindings should use the properties
defined below where they are applicable; not all of these properties
are relevant or useful for all hardware or binding structures. Each
individual binding document should state which of these generic
properties, if any, are used, and the structure of the DT nodes that
contain these properties.

This is based on Linux, documentation:
https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml

Add pincfg-node.yaml properties and refactor atmel,sam-pinctrl.yaml
to use these properties.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

galak commented on dts/bindings/pinctrl/atmel,sam-pinctrl.yaml:
```diff
@@ -16,22 +16,6 @@ properties:
       const: 1
 
 child-binding:
-    description: atmel pins
```
Are you able to leave `description:` here ?

nandojve replied:
Fixed!

galak commented:
@erwango FYI on dts/bindings/pinctrl/pincfg-node.yaml in this PR.  We might want to pull that out into its own PR so it can be used for the ST pinctrl as well.

nandojve committed:
soc: arm: cypress: psoc6: Add Cortex-M0+ mux support

PSoC-6 SoC needs that user define the nvic interrupt number
to bind with the peripheral interrupt line for the Cortex-M0+
CPU.  It uses a multiplex before any NVIC interrupt line.
The minimal support was added to allow interrupts be available
for both CPUs.

Note: The PSoC-6 SoC allows that both CPUs receive the same
interrupt.  A tipical use is GPIO interrupt handle and user is
responsable to define interrupt line, priority and take care
of enable same peripheral instance on both CPUs only when
appropriated.

Signed-off-by: Gerson Fernando Budke <gerson.budke@atl-electronics.com>

erwango commented:
> @erwango FYI on dts/bindings/pinctrl/pincfg-node.yaml in this PR. We might want to pull that out into its own PR so it can be used for the ST pinctrl as well.

I agree reusing the same binding makes sense, and I don't see any reason why it shouldn't work. Though, a mentioned:
```
  Each individual binding document
  should state which of these generic properties, if any, are used, and the
  structure of the DT nodes that contain these properties.
```
So:
- we need to overwrite each of the property used anyway
- this makes a number of non supported properties available to the user (which would compile but get no effect), while today we get an error when a non supported property is set:
```
devicetree error: 'skew-delay' appears in /soc/pin-controller@48000000/usart1_tx_pb6 in disco_l475_iot1.dts.pre.tmp, but is not declared in 'properties:' in /local/mcu/zephyrproject/zephyr/dts/bindings/pinctrl/st,stm32-pinctrl.yaml
```

galak reviewed (changes requested):
Can you pull the `include: sys: util: Move macros from util to util_macro` into its own PR so the guys that maintain that can review and comment.

galak reviewed (commented):
Should 'soc: arm: cypress: psoc6: Add Cortex-M0+ mux support' be part of the PR?

gmarull commented on dts/bindings/pinctrl/cypress,psoc6-pinctrl.yaml:
```diff
@@ -0,0 +1,23 @@
+    description: cypress pins
+    properties:
+      "cypress,pins":
+        type: phandle-array
```
Can you provide a usage example of the pinctrl bindings (board)?

nandojve replied:
Yes, to have something that works need solve dependencies first: dt-util.h, IRQ and GPIO.

gmarull reviewed (commented):
Added some initial comments.

nandojve commented on dts/bindings/pinctrl/atmel,sam-pinctrl.yaml:
```diff
@@ -20,18 +20,3 @@ child-binding:
-      drive-open-drain:
-        required: false
-        type: boolean
-        description: drive with open drain
```
It seems that somehow we need inform user about what props are available.

1) Is there any consensus how to do it?
2) I noted that **slew-rate** type on ST's files are string and at https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml defines as uint32:
```
  slew-rate:
    $ref: /schemas/types.yaml#/definitions/uint32
    description: set the slew rate

```
https://github.com/zephyrproject-rtos/zephyr/blob/f219a982ffda7496549fab82ae864afffb1303b9/dts/bindings/pinctrl/st%2Cstm32-pinctrl.yaml#L76-L87
It will be better if you already have the final solution on master, so I can rebase and apply [1].

CC: @erwango , @galak

nandojve commented on dts/arm/cypress/pinctrl_cypress_psoc6.h:
```diff
@@ -0,0 +1,100 @@
+ * NODE = p<port>_<pin>_<inst>_<signal>
+ *
+ * NODE: NODE {
+ *	cypress,pins = < &p<port> <port> <pin> HSIOM_SEL_<hsiom> >;
```
I think it already is at end:

This is the structure that will handle:
```
struct soc_gpio_pin {
	GPIO_PRT_Type *regs; /** pointer to registers of the GPIO controller */
	uint32_t pinum;      /** pin number */
	uint32_t flags;      /** pin flags/attributes */
};
```

The **soc/arm/cypress/common/cypress_psoc6_dt.h** have a macro **CY_PSOC6_DT_PIN** that will collect from DT and create above structure at driver level at build time.

The DT_CYPRESS_HSIOM is a variable length argument list that accept flags:
First 5 parameters are mandatory, than flags are added one by one with commas.
DT_CYPRESS_HSIOM(uart5,  rx,  5, 0, act_6);
DT_CYPRESS_HSIOM(uart5,  rx,  5, 0, act_6, bias-pull-up);
DT_CYPRESS_HSIOM(uart5,  rx,  5, 0, act_6, bias-pull-up, input-enable [, etc]);

Above all entries are valid and create a node like:

```
 p5_0_uart5_rx: p5_0_uart5_rx {
    cypress,pins = <&p5 0x5 0x0 0x12 >;
    bias-pull-up;
    input-enable;
}
```

The CY_PSOC6_DT_PIN will combine 0x12 (function) + bias-pull-up + input-enable as flags. From &p5 we get reg addr and 0x0 is the pin itself at p5. The remaining 0x5 is the uart instance number and doesn't have nothing to do with pinctrl and probably will be dropped. It was necessary to solve the chicken and egg problem when platform doesn't have nothing.

nandojve commented on dts/arm/cypress/psoc6-pinctrl.dtsi:
```diff
@@ -0,0 +1,19 @@
+			DT_CYPRESS_HSIOM(uart5,  rx,  5, 0, act_6, input-enable);
+			DT_CYPRESS_HSIOM(uart5,  tx,  5, 1, act_6, drive-push-pull);
+			DT_CYPRESS_HSIOM(uart6,  rx, 13, 0, act_6, input-enable);
+			DT_CYPRESS_HSIOM(uart6,  tx, 13, 1, act_6, drive-push-pull);
```
I don't have intention to move this to HAL and mix Zephyr with Cypress.

erwango replied:
It doesn't have to be in PDL directly. It could be inside a dts/ at cypress module root.

nandojve commented on dts/bindings/gpio/cypress,psoc6-gpio.yaml:
```diff
@@ -0,0 +1,36 @@
+cypress,pin-cells:
+  - port
+  - pin
+  - hsiom
```
It is not right, To solve IRQ, GPIO, PINCTRL and noted that the right sequence is:

dt-bindings/dt-util.h: #28779
IRQ: #29459
GPIO: working on
PINCTRL: This PR

nandojve commented on soc/arm/cypress/psoc6/soc.h:
```diff
@@ -14,14 +15,17 @@
- */
-#include <kernel_includes.h>
+#include "../common/soc_gpio.h"
+#include "../common/cypress_psoc6_dt.h"
```
Cypress uses PDL with PSoC 4 and PSoC 6. PSoC-6 is an dual core asymmetric (Cortex-M0+/M4) and PSoC-4 is a Cortex-M0+. It will be natural share some code once both shares same library.
